### PR TITLE
Cleanup ORCA compiler flags

### PR DESCRIPTION
**Pull request by hlinnaka**
Since https://github.com/greenplum-db/gporca/pull/137, ORCA contains a config.h file, which simplifies the build process because we don't need to derive GPOS_DEBUG and other such flags ourselves.

We no longer try to guess whether we're making a debug or release build ourselves in gpopt.mk. Rely on the global compiler flags derived by configure, also for the C++ code.

### CODE REVIEW

**Comment by hlinnaka** [comment]
This method of deriving CXXFLAGS from CFLAGS was previously discussed at https://github.com/greenplum-db/gpdb/pull/1200, but we dropped the ball then.

**Comment by vraghavan78** [comment]
@d @oarap @hsyuan please review this PR

**Comment by oarap** [comment]
@d and I are on it

File: src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp [hunk 1]
```diff
@@ -486,7 +486,7 @@ CTranslatorQueryToDXL::PdxlnFromQueryInternal()
 	CDXLNode *pdxlnCTEAnchorTop = NULL;
 	CDXLNode *pdxlnCTEAnchorBottom = NULL;
 	ConstructCTEAnchors(m_pdrgpdxlnCTE, &pdxlnCTEAnchorTop, &pdxlnCTEAnchorBottom);
-	GPOS_ASSERT_IMP(0 < m_pdrgpdxlnCTE->UlSafeLength(),
```
Comment (vraghavan78) [inline src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp]: UlSafeLength() does exactly that.

Comment (vraghavan78) [inline src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp]: UlSafeLength does the null check.

File: src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp [hunk 2]
```diff
@@ -5862,7 +5860,7 @@ CTranslatorDXLToPlStmt::PdistrpolicyFromCtas
 {
 	DrgPul *pdrgpulDistrCols = pdxlop->PdrgpulDistr();
 
-	const ULONG ulDistrCols = pdrgpulDistrCols->UlSafeLength();
```
Comment (vraghavan78) [inline src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp]: UlSafeLength does the null check. so same logic

**Comment by hlinnaka** [comment]
On 19 January 2017 21:25:50 EET, Venkatesh Raghavan <notifications@github.com> wrote:
>vraghavan78 commented on this pull request.
>
>
>
>> @@ -486,7 +486,7 @@ CTranslatorQueryToDXL::PdxlnFromQueryInternal()
> 	CDXLNode *pdxlnCTEAnchorTop = NULL;
> 	CDXLNode *pdxlnCTEAnchorBottom = NULL;
>	ConstructCTEAnchors(m_pdrgpdxlnCTE, &pdxlnCTEAnchorTop,
>&pdxlnCTEAnchorBottom);
>-	GPOS_ASSERT_IMP(0 < m_pdrgpdxlnCTE->UlSafeLength(),
>
>UlSafeLength does the null check.

Right. The point of this change is to stop using UlSafeLength(), because it gives compiler warnings. (See commit message.)

- Heikki

**Comment by vraghavan78** [comment]
Thanks for the explanation about UlSafeLength()

**Comment by d** [comment]
Mostly looks good to us except:

0. Subtly different behavior: we seem to be losing a `-g` and `-g3` on `src/backend/gpopt`
0. Subtly different behavior: after this patch, if we run `./configure` we don't get a `-g` flag on the translator either

Are these behaviors intentional?

**Comment by hlinnaka** [comment]
On 01/20/2017 02:19 AM, Jesse Zhang wrote:
> Mostly looks good to us except:
>
> 0. Subtly different behavior: we seem to be losing a `-g` and `-g3` on `src/backend/gpopt`
> 0. Subtly different behavior: after this patch, if we run `./configure` we don't get a `-g` flag on the translator either
>
> Are these behaviors intentional?

Yeah, those flags are now set the same as in the rest of the system. We 
don't use -g3 anywhere else, and we use -g only when the --enable-debug 
configure option is used.

IMO it's always a good idea to use --enable-debug, to get -g, even in 
production builds, because it doesn't really cost anything else than 
larger binaries. But in any case, there's no reason for 
src/backend/gpopt to be different from the rest of the server. I don't 
know what the difference between -g3 and -g is.

You can always use "configure CFLAGS=-g3" locally, if you want that, though.

- Heikki

**Comment by danielgustafsson** [comment]
`-g3` is a GCC option to store all macro definitions (and possible something more which I can't remember right now) in the produced binary.

**Comment by danielgustafsson** [comment]
LGTM

**Comment by hlinnaka** [comment]
Pushed, thanks everyone for the review.

**Pull request closed without merging**